File: backend/utils/report_helper.py
import json, os

CONFIG_PATH = os.path.join(os.path.dirname(__file__), '../../shared/report_config.json')

def load_config():
    with open(CONFIG_PATH) as f:
        return json.load(f)

def get_columns(report_name, subkey="columns"):
    return load_config()[report_name][subkey]

def fmt_val(value, col_type):
    if col_type == "qntl":
        return round((value or 0) / 100, 2)
    elif col_type == "integer":
        return int(value or 0)
    elif col_type == "number":
        return value or 0
    return value or ""

def get_entry_row(entry, columns):
    return [fmt_val(entry.get(col["field"], 0), col["type"]) for col in columns]

def get_total_row(totals, columns):
    return [fmt_val(totals.get(col["total_key"], 0), col["type"]) if col.get("show_total") else None for col in columns]

def get_excel_headers(columns):
    return [c["header"] for c in columns]

def get_pdf_headers(columns):
    return [c["pdf_header"] for c in columns]

def get_excel_widths(columns):
    return [c["width_excel"] for c in columns]

def get_pdf_widths_mm(columns):
    return [c["width_pdf_mm"] for c in columns]

def col_count(columns):
    return len(columns)

File: desktop-app/shared/report_helper.js
const fs = require('fs');
const path = require('path');

let _config = null;
function loadConfig() {
    if (!_config) _config = JSON.parse(fs.readFileSync(path.join(__dirname, 'report_config.json')));
    return _config;
}

function getColumns(reportName, subkey) {
    subkey = subkey || 'columns';
    return loadConfig()[reportName][subkey];
}

function fmtVal(value, type) {
    if (type === 'qntl') return Math.round((value || 0) / 100 * 100) / 100;
    if (type === 'integer') return Math.round(value || 0);
    if (type === 'number') return value || 0;
    return value || '';
}

function getEntryRow(entry, columns) {
    return columns.map(col => fmtVal(entry[col.field], col.type));
}

function getTotalRow(totals, columns) {
    return columns.map(col => col.show_total && col.total_key ? fmtVal(totals[col.total_key], col.type) : null);
}

function getExcelHeaders(columns) { return columns.map(c => c.header); }
function getPdfHeaders(columns) { return columns.map(c => c.pdf_header); }
function getExcelWidths(columns) { return columns.map(c => c.width_excel); }
function getPdfWidthsMm(columns) { return columns.map(c => c.width_pdf_mm); }
function colCount(columns) { return columns.length; }

function getTotalKeys(columns) {
    return columns.filter(c => c.show_total && c.total_key).map(c => c.total_key);
}

module.exports = { getColumns, fmtVal, getEntryRow, getTotalRow, getExcelHeaders, getPdfHeaders, getExcelWidths, getPdfWidthsMm, colCount, getTotalKeys };
